packages/react-aria-components/src/Separator.tsx

```diff
@@ -11,17 +11,33 @@
  */
 
 import {SeparatorProps as AriaSeparatorProps, useSeparator} from 'react-aria';
+import {BaseCollection, CollectionNode, createLeafComponent} from '@react-aria/collections';
 import {ContextValue, SlotProps, StyleProps, useContextProps} from './utils';
-import {createLeafComponent} from '@react-aria/collections';
 import {filterDOMProps, mergeProps} from '@react-aria/utils';
-import {GlobalDOMAttributes} from '@react-types/shared';
+import {GlobalDOMAttributes, Key} from '@react-types/shared';
 import React, {createContext, ElementType, ForwardedRef} from 'react';
 
 export interface SeparatorProps extends AriaSeparatorProps, StyleProps, SlotProps, GlobalDOMAttributes<HTMLElement> {}
 
 export const SeparatorContext = createContext<ContextValue<SeparatorProps, HTMLElement>>({});
 
-export const Separator = /*#__PURE__*/ createLeafComponent('separator', function Separator(props: SeparatorProps, ref: ForwardedRef<HTMLElement>) {
+export class SeparatorNode extends CollectionNode<any> {
+  static readonly type = 'separator';
+
+  constructor(key: Key) {
+    super(SeparatorNode.type, key);
+  }
+
+  filter(collection: BaseCollection<any>, newCollection: BaseCollection<any>): CollectionNode<any> | null {
+    if (newCollection.getItem(this.prevKey!)) {
+      return this.clone();
+    }
```

Comment:
if you have parentKey or nextKey, could you decide to return null here? saw you had a question of if you could move that logic in here

Reply:
I guess I could do the callback, trying to think about how filterChildren would go about calling that callback before it returns the first/last node (maybe filter would return a callback alongside the CollectionNode)? Still feels easier to handle this logic in filterChildren though...

As for the css approach, that would mean the user would need to implement that no? Wouldn't be too hard I guess, but kinda annoying that they would need to know to do so.

Reply:
yeah, not ideal
It would be nice if we could queue additional checks, will think on it some more